feat(runner): Integrate Test262 execution with wptrunner and add smoke tests [pt 3/5]

jcscottiii opened the pull request:
### Summary
Integrates the TC39 Test262 test suite natively into the Web Platform Tests infrastructure. This enables automated, cross-browser ECMAScript conformance testing directly within `wptrunner` and standard WPT dashboards.

### Architecture
Test262 tests are executed via a custom Window Wrapper implementation that maps the standard Test262 host API into WPT's `testharness.js`. 

* **Wrapper Generation:** `tools/serve/serve.py` intercepts `.test262.html` requests and dynamically generates a document that injects `testharness.js`, the `$262` host provider, and Test262's standard evaluation fixtures.
* **Execution Boundary:** Tests are isolated within an iframe child context to prevent global scope contamination and enforce strict mode expectations.
* **Communication Protocol:** A `postMessage` bridge (`test262-reporter.js` -> `test262-harness-bridge.js`) synchronizes child-frame execution states. The adapter emits structured JSON payloads (`{type: 'complete'}`, `{type: 'fail'}`, `{type: 'error'}`) that trigger `t.done()`, `assert_unreached()`, and generic unhandled errors in the parent. This natively delegates all timeout and lifecycle management directly to `testharness.js`.

### Supported Features
* **Negative Tests (`negative`):** Safely traps expected parsing and runtime evaluation exceptions and translates them to successful Web Platform Tests assertions.
* **Asynchronous Tests (`async`):** Configures `$DONE` and `print()` bindings to fulfill Test262 async constraints before halting the overarching WPT test pipeline.
* **Modules (`module`):** Injects dynamic module loading wrappers, directly escalating syntax parsing and import boundary rejections into explicit WPT subtest failures.
* **Strict mode (`onlyStrict`):** Supports execution isolation for strictly defined environment iterations.

jcscottiii committed:
feat(runner): Integrate Test262 execution with wptrunner and add smoke tests

This commit completes the core integration of Test262 into the WPT
runner (`wptrunner`), enabling the execution of Test262 JavaScript
conformance tests within the browser environment. This work builds upon
the manifest tooling (PR 1 #56840) and server-side handling (PR 2 #56841) to provide
a full end-to-end testing solution for Test262.

Key components introduced and integrated:
- **Client-Side Harness:** New JavaScript files (`harness-adapter.js`,
  `testharness-client.js`, `testharness.js`) under `resources/test262/`
  provide the necessary environment for Test262 tests to execute in an
  `iframe` and report results back to the main WPT testharness.
- **Test262 Assertion Libraries (Vendored):** For the purpose of enabling these
  initial smoke tests, the standard Test262 assertion libraries (`assert.js`
  and `sta.js`) are directly placed into `third_party/test262/harness/`.
  This provides the immediate dependencies required by the smoke tests.
  An autoamted vendoring solution, including version tracking via
  `vendored.toml`, will be implemented in a subsequent PR focused on
  automated updates, as described in the RFC.
- **Wptrunner Integration:** Modifications to `tools/wptrunner/wpttest.py`
  and `tools/wptrunner/browsers/*.py` enable `wptrunner` to recognize the
  `test262` test type and execute it using existing testharness executors.
- **Smoke Tests and Metadata:** A suite of basic Test262 smoke tests
  (`infrastructure/test262/`) and their corresponding expected results
  (`infrastructure/metadata/infrastructure/test262/`) are added to verify
  the correct functioning of the entire integration.
- **Linting Exemption:** `lint.ignore` is updated to exclude the vendored
  Test262 harness files from linting, respecting their upstream style.

This integration allows browser vendors to run Test262 directly with `wptrunner`,
streamlining conformance testing efforts as specified in the RFC:
https://github.com/web-platform-tests/rfcs/pull/229

This commit is the third in a series of smaller PRs split from the larger,
original implementation in https://github.com/web-platform-tests/wpt/pull/55997.

DanielRyanSmith reviewed (commented):
This suggestion came from a Gemini code review:

### **Harness Adapter Robustness**

**Focus:** `resources/test262/harness-adapter.js`
**Critique:** Bridging Test262's global-polluting API (e.g., `$DONE`, `Test262Error`) to the structured `testharness.js` API is error-prone.

* **Exception Handling:** Ensure the adapter explicitly wraps the execution of the Test262 test code in a `try/catch` block (or uses `window.onerror`). Test262 tests often throw bare strings or non-Error objects. The adapter **must** normalize these into meaningful `testharness` failure messages, otherwise, the runner might just report a generic "Script error." or timeout.
* **$DONE Semantics:** Verify that calling `$DONE()` (without arguments) signals a **PASS**, and `$DONE(error)` signals a **FAIL**. A common bug is treating any call to `$DONE` as a completion without checking the argument.

o- commented on resources/test262/harness-adapter.js:
```diff
@@ -0,0 +1,173 @@
+      }
+    },
+    detachArrayBuffer: function(buffer) {
+      if (typeof postMessage !== 'function') {
```
This happens if the environment does not support workers. I guess on the web that should never happen (?), so we can probably remove that case.

o- commented on resources/test262/harness-adapter.js:
```diff
@@ -0,0 +1,173 @@
+    createRealm: function() {
+      var iframe = global.document.createElement('iframe');
+      iframe.style.cssText = 'display: none';
+      iframe.src = '';  // iframeSrc;
```
Since I have not written this part I am not fully sure. What it does is set the src to the same document as the parent page. This does not look like what we want. @jcscottiii can we try what happens if we just remove it?

jcscottiii committed:
address feedback

jcscottiii committed:
address more feedback

jcscottiii committed:
address feedback

DanielRyanSmith reviewed (approved):
## I'm going to approve this, but a typo was found that needs to be fixed:

**Critical Typo in Safari Configuration**

In `tools/wptrunner/wptrunner/browsers/safari.py`, the `crashtest` executor mapping was accidentally renamed to `crashtag`. This will break the execution of crash tests in Safari.

```python
@@ -22,7 +22,8 @@
                  "executor": {"testharness": "WebDriverTestharnessExecutor",
                               "reftest": "WebDriverRefTestExecutor",
                               "wdspec": "WdspecExecutor",
-                              "crashtest": "WebDriverCrashtestExecutor"},
+                              "crashtag": "WebDriverCrashtestExecutor",
+                              "test262": "WebDriverTestharnessExecutor"},
```

**Recommendation:** Revert `crashtag` back to `crashtest`.

Other than this issue, the core integration of Test262, including the iframe realm isolation, `$262` agent logic, and runner mappings, looks well-implemented and structurally sound.

jcscottiii committed:
address feedback

jcscottiii committed:
Apply suggestion from @jcscottiii

jcscottiii committed:
fix lint

jcscottiii committed:
move need third_party lint because other doesn't work

jcscottiii committed:
fix lint

jcscottiii committed:
Refactor Test262 integration to remove custom custom harness bridge
This commit addresses reviewer feedback (jgraham) regarding the Test262
integration "reimplementing a subset of testharness.js". The architecture
has been drastically simplified by eliminating the custom `postMessage`
protocol between the Test262 iframe and the parent WPT test wrapper.
Changes include:
- Deleted `test262-harness-bridge.js` completely.
- Updated the wrapper `<script>` generated by `tools/serve/serve.py` to
  expose a native `window.test262HarnessDone(status, error)` hook directly
  in the parent's `testharness.js` execution environment.
- Refactored `test262-reporter.js` to natively dispatch test completions
  and harness errors to this synchronous hook instead of manually wrapping
  results in `{type: "complete"}` message payloads.
- Ensured strict "Realm Isolation" is maintained for the `test262` tests,
  as required by `INTERPRETING.md`, by keeping all WPT globals cleanly
  isolated in the parent window.
- Updated `tools/serve/test_serve.py` assertions to reflect the removed
  bridge dependency.
- Leveraged `setTimeout` to safely elevate test harness crashes into the
  parent's macro-task queue, ensuring infrastructure `.ini` expectations
  (like timeouts and parsing ERRORs) correctly fail the WPT harness.

jcscottiii committed:
Update tools/serve/serve.py

jcscottiii committed:
update to docs from last commits

jcscottiii committed:
touch ups

jcscottiii committed:
nit

DanielRyanSmith reviewed (approved):
This LGTM to me, and looks ready to merge. 🙂 No more suggestions on my end.

@jgraham anymore feedback?

jgraham commented on resources/test262/test262-provider.js:
```diff
@@ -0,0 +1,214 @@
+     * Triggers garbage collection if supported by the host.
+     */
+    gc: function() {
+      throw new Error('Test262 Host API: gc() not supported');
```
`TestUtils.gc()` is expected to work in wpt, we can use that (and maybe catch the error if it actually doesn't work).

jcscottiii replied:
I updated it to use TestUtils.gc() and added a try-catch block to handle cases where it might fail or not be supported.

jgraham commented on resources/test262/test262-provider.js:
```diff
@@ -0,0 +1,214 @@
+  };
+
+  // Ensure print is available in this realm if it's available in the parent
+  if (typeof global.print === 'undefined' && typeof window.print === 'function') {
```
`window.print` is generally a function in DOM windows (it tries to print the page), so this seems sketchy, in that it will only be undefined in workers, but maybe it's unavoidable?

jcscottiii replied:
Now unconditionally overriding global.print = window.print. Let me know what you think.

jgraham commented on resources/test262/test262-reporter.js:
```diff
@@ -0,0 +1,142 @@
+            return;
+        }
+
+        if (test_finished) return;
```
Braces, please

jcscottiii replied:
Done!

jgraham commented on resources/test262/test262-provider.js:
```diff
@@ -0,0 +1,214 @@
+
+  global.$DONE = function(err) {
+    if (err) {
+      if (typeof global.print === 'function') {
```
There seem to be silent failure cases here where `global.print` isn't defined. Is that actually required or is that code supposed to be unreachable? If it's required I think we need a comment, otherwise some kind of error (also the conditionals are arguably the wrong way around; we could test for `global.print` once on the outside of the `if (err)` condition).

jcscottiii replied:
I removed the redundant $DONE implementation since Test262's doneprintHandle.js is automatically injected for async tests and provides it. I left a placeholder $DONE for non-async tests that throws a clear error if called accidentally.

jgraham commented on resources/test262/test262-reporter.js:
```diff
@@ -0,0 +1,142 @@
+    window.test262Setup = function() {
+    };
+
+    window.test262IsAsync = function(b) {
```
Is there a reason for the cryptic one-letter argument names here?

jcscottiii replied:
I renamed the cryptic one-letter arguments to clearer, more descriptive names.

jgraham commented on resources/test262/test262-reporter.js:
```diff
@@ -0,0 +1,142 @@
+    function on_error(event) {
+        // This hack ensures that errors thrown inside of a $262.evalScript get
+        // rethrown in the correct place.
+        if (event.error && String_prototype_indexOf.call(event.error.message, "Failed to execute 'appendChild' on 'Node'") === 0) {
```
This confuses me. If we're trying to catch a specific error then I don't think we can use the message like this because it's browser specific.

jcscottiii replied:
Good call out. Yeah this is a little fragile.
Replaced with a state flag (__test262_evalScript_active_) to identify errors during script evaluation. The idea is that this flag is only true during the synchronous execution of appendChild(script). This allows us to reliably capture any error thrown by the evaluated script and rethrow it synchronously, without relying on fragile error message strings. 

But open to other ideas!

jgraham commented on resources/test262/test262-reporter.js:
```diff
@@ -0,0 +1,142 @@
+         */
+        let errorMatches = false;
+        if (expectedType && event.error) {
+            if (String_prototype_indexOf.call(event.error.toString(), expectedType) === 0 ||
```
So, if I understand https://tc39.es/ecma262//#sec-error.prototype.tostring correctly, this works because the first part of the `toString` value is always a string representation of the `name` property of the Error object (or its prototype). Is there a specific advantage of doing it this way vs just reading that property directly?

jcscottiii replied:
A little pivot from what I originally had. I updated the error matching logic to use event.error.constructor.name directly. This aligns with Test262 INTERPRETING.md which explicitly states that we must match the name of the exception's constructor.

jgraham commented on tools/serve/serve.py:
```diff
@@ -333,8 +334,27 @@ class Test262WindowHandler(HtmlWrapperHandler):
+<script>
+setup({explicit_done: true});
+const t = async_test(document.title);
+window.test262HarnessDone = function(status, message) {
```
Why not make this
```js
window.test262HarnessDone = t.step_func(function(status, message) {
[…]
});
```
rather than wrapping each individual branch in a step? In fact `step_func_done()` is probably what you want, given that the test is always done after this step.

jcscottiii replied:
Done! Thanks!

jgraham commented on tools/serve/serve.py:
```diff
@@ -333,8 +334,27 @@ class Test262WindowHandler(HtmlWrapperHandler):
+        t.done();
+        done();
+    } else if (status === 2) {
+        t.step(() => assert_unreached(message || "Harness Error"));
```
Why not just throw the Error here?

jcscottiii replied:
Done. Thanks!

jgraham commented on tools/serve/serve.py:
```diff
@@ -333,8 +334,27 @@ class Test262WindowHandler(HtmlWrapperHandler):
+<script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
+<script>
+setup({explicit_done: true});
```
Is `explicit_done` actually needed here? The test is created before the load is complete, so there's no problem where we're racing creating the test with the harness trying to complete. And the fact that you always call `done()` right after `t.done()` is suspicious.

jcscottiii replied:
Good catch!  Removed.

jgraham commented on tools/serve/serve.py:
```diff
@@ -359,40 +379,113 @@ class Test262WindowTestBaseHandler(HtmlWrapperHandler):
+            return """<script>test262IsAsync(true)</script>"""
         return None
 
+    def handle_request(self, request, response):
```
I don't think we should be replacing the entire base class function here. At worst you should be able to override `_get_meta` and `_get_script` with the functionality you want. AFAICT the only real additional difficulty is that you want the additional `done_script` parameter for async tests. That seems like something that could be incorporated into the base class if necessary.

jcscottiii replied:
Done. Using _get_meta and _get_script  now.

jcscottiii committed:
start to address feedback

jcscottiii committed:
finish feedback

jcscottiii committed:
tidy up

jcscottiii committed:
doc touchups

jcscottiii committed:
add missing support for phases

jcscottiii commented:
@jgraham This is now updated to address the feedback. Additionally, I noticed the original code was missing support for validating negative test phases (parse, early, resolution), so I have implemented that and added smoke tests to verify it.

jgraham commented on resources/test262/test262-provider.js:
```diff
@@ -58,7 +62,11 @@ function installAPI(global) {
     gc: function() {
-      throw new Error('Test262 Host API: gc() not supported');
+      try {
+        TestUtils.gc();
```
Although thinking about it, this is async, so it's not guaranteed that it's complete by the time that the `gc()` method returns. I'm not sure what to do about that; ideally the test262 interface would change, but if not we could consider adding a `TestUtils.blocking_gc()` method that would allow this to work. At least in Gecko I think that's implementable.

jcscottiii replied:
Filed an issue on Test262: https://github.com/tc39/test262/issues/5029

And a back up [issue](https://github.com/web-platform-tests/wpt/issues/59239) in case test262 discussion concludes that it should be done in WPT.

jgraham commented on resources/test262/test262-reporter.js:
```diff
@@ -89,32 +110,44 @@
+                message = "OK";
+            } else if (expectedPhase === PHASE_RUNTIME) {
+                status = 0;
+                message = "OK";
```
These two blocks look identical?

jcscottiii replied:
Good catch. I have combined those identical branches in on_error.

jgraham commented on resources/test262/test262-reporter.js:
```diff
@@ -33,36 +34,56 @@
 
-    window.test262NegativeType = function(t) {
-        expectedType = t;
+    window.test262NegativeType = function(type) {
```
FWIW it seems like we always call `test262NegativeType` and `test262NegativePhase` together, so I wonder if one function with two arguments would be better?

jcscottiii replied:
Makes sense. I've consolidated test262NegativeType and test262NegativePhase into a single test262Negative(type, phase) call.

jgraham reviewed (approved):
I think some small changes are required, and I'm not totally confident I fully understand all the details of the test protocol, but lets mark this as approved assuming you address the issues; the next stage is probably to land this and see how well it works.

jcscottiii committed:
final touchups

trflynn89 commented on resources/test262/test262-provider.js:
```diff
@@ -0,0 +1,220 @@
+    /**
+     * Detaches an ArrayBuffer.
+     */
+    detachArrayBuffer: function(buffer) {
+      // Transfer and thus detach the buffer.
+      postMessage(null, '*', [buffer]);
+    },
```
Just wanted to point out that this doesn't work on tests which call `$DETACHBUFFER` multiple times. For example:

https://github.com/web-platform-tests/wpt/blob/f2535c6918/third_party/test262/test/built-ins/TypedArray/prototype/sort/sort-tonumber.js

https://wpt.fyi/results/third_party/test262/test/built-ins/TypedArray/prototype/sort/sort-tonumber.test262.html?product=ladybird&product=chrome%5Bstable%5D&product=firefox%5Bstable%5D&product=safari%5Bstable%5D

The array buffer is detached inside a `sort` invocation and thus called multiple times. So every engine throws an exception on the second invocation.
